### PR TITLE
Update Core Helpers

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Altair Work

**What does this PR do? Why is it needed?**

- [x] Removes the active balance cache from our committee cache.
- [x] Adds in Sync Committee helpers.
- [x] Adds in relevant Sync Period and Subcommittee helpers.
- [x] Adds in relevant tests for methods.

**Which issues(s) does this PR fix?**

Part of #8638 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/core/helpers/slot_epoch.go [hunk 1]
```diff
@@ -208,8 +208,8 @@ func PrevSlot(slot types.Slot) types.Slot {
 // Spec code:
 // def compute_sync_committee_period(epoch: Epoch) -> uint64:
 //    return epoch // EPOCHS_PER_SYNC_COMMITTEE_PERIOD
-func SyncCommitteePeriod(e types.Epoch) uint64 {
-	return uint64(e / params.BeaconConfig().EpochsPerSyncCommitteePeriod)
+func SyncCommitteePeriod(epoch types.Epoch) uint64 {
```
Comment (terencechain) [inline beacon-chain/core/helpers/slot_epoch.go:211]: I think we can revert this as well

File: beacon-chain/core/helpers/committee.go [hunk 2]
```diff
@@ -271,14 +273,8 @@ func VerifyAttestationBitfieldLengths(state state.ReadOnlyBeaconState, att *ethp
...
-// ShuffledIndices uses input beacon state and returns the shuffled indices of the input epoch,
-// the shuffled indices then can be used to break up into committees.
-func ShuffledIndices(s state.ReadOnlyBeaconState, epoch types.Epoch) ([]types.ValidatorIndex, error) {
-	seed, err := Seed(s, epoch, params.BeaconConfig().DomainBeaconAttester)
-	if err != nil {
-		return nil, errors.Wrapf(err, "could not get seed for epoch %d", epoch)
-	}
-
+// Returns the active indices of the validators in input `state` and during input `epoch`.
+func activeIndices(s state.ReadOnlyBeaconState, epoch types.Epoch) ([]types.ValidatorIndex, error) {
```
Comment (terencechain) [inline beacon-chain/core/helpers/committee.go:277]: We should revert these changes as well

File: beacon-chain/core/helpers/committee.go [hunk 3]
```diff
@@ -383,10 +385,200 @@ func UpdateProposerIndicesInCache(state state.ReadOnlyBeaconState) error {
...
 func ClearCache() {
 	committeeCache = cache.NewCommitteesCache()
 	proposerIndicesCache = cache.NewProposerIndicesCache()
+	syncCommitteeCache = cache.NewSyncCommittee()
+}
+
+// IsCurrentPeriodSyncCommittee returns true if the input validator index belongs in the current period sync committee
+// along with the sync committee root.
+// 1.) Checks if the public key exists in the sync committee cache
+// 2.) If 1 fails, checks if the public key exists in the input current sync committee object
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/committee.go:398]: > along with the sync committee root

What does this mean? Was this supposed to return the root?

File: beacon-chain/core/helpers/committee.go [hunk 4]
```diff
@@ -383,10 +385,200 @@ func UpdateProposerIndicesInCache(state state.ReadOnlyBeaconState) error {
...
+		return len(findSubCommitteeIndices(val.PublicKey, committee.Pubkeys)) > 0, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return len(indices) > 0, nil
+}
+
+// IsNextPeriodSyncCommittee returns true if the input validator index belongs in the next period sync committee
+// along with the sync period boundary root.
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/committee.go:433]: Here too. Is this supposed to return the root?

File: beacon-chain/core/helpers/committee.go [hunk 5]
```diff
@@ -383,10 +385,200 @@ func UpdateProposerIndicesInCache(state state.ReadOnlyBeaconState) error {
...
+		return findSubCommitteeIndices(val.PublicKey, committee.Pubkeys), nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	return indices, nil
+}
+
+// UpdateSyncCommitteeCache updates sync committee cache.
+// It uses `state`'s latest block header root as key. To avoid miss usage, it disallows
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/committee.go:523]:
```suggestion
// It uses `state`'s latest block header root as key. To avoid misuse, it disallows
```

File: beacon-chain/core/helpers/committee.go [hunk 6]
```diff
@@ -383,10 +385,200 @@ func UpdateProposerIndicesInCache(state state.ReadOnlyBeaconState) error {
...
+			return false, nil
+		}
+		committee, err := st.CurrentSyncCommittee()
+		if err != nil {
+			return false, err
+		}
+
+		// Fill in the cache on miss.
+		go func() {
+			if err := syncCommitteeCache.UpdatePositionsInCommittee(bytesutil.ToBytes32(root), st); err != nil {
```
Comment (rauljordan) [inline beacon-chain/core/helpers/committee.go:419]: Should this be a blocking operation? scared of having these side-effects that are asynchronous and we cannot know when they will complete

File: beacon-chain/core/helpers/committee.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/core/helpers/committee.go:398]: Bad sentence. It means `using the sync period root as key`. Will fix

File: beacon-chain/core/helpers/committee.go (hunk 6 above)
Comment (terencechain) [reply beacon-chain/core/helpers/committee.go:419]: I dont have a strong opinion here. `UpdatePositionsInCommittee` is really fast (a few ms). Most of our code updates cache in the back ground as non-blocking so this followed similar pattern

Can remove if anyone feels that's safer

**Review by terencechain — APPROVED** [review]
